From: cjao <casey.jao@gmail.com>
Date: Fri, 17 Jul 2020 07:00:43 -0400
Subject: Add some print capabilities to the print dialog. (#320)

* Add some print capabilities to the print dialog.

Use the same capabilities as in gtkprintoperation-unix.c.

* Fix a race condition in print.c.

The temporary file for print preview was previously being removed
before evince-previewer finished launching.

Origin: upstream, 1.7.2, commit:7815428821a9c09626433e5f2a10ed67c506ddc8
---
 src/print.c | 28 ++++++++++++++++++++++------
 1 file changed, 22 insertions(+), 6 deletions(-)

diff --git a/src/print.c b/src/print.c
index ce4d435..df6a357 100644
--- a/src/print.c
+++ b/src/print.c
@@ -303,7 +303,6 @@ print_file (int fd,
                               NULL,
                               error) == -1)
     return FALSE;
-
   if (job)
     {
       if (!gtk_print_job_set_source_file (job, filename, error))
@@ -314,13 +313,17 @@ print_file (int fd,
   else
     {
       if (!launch_preview (filename, title, settings, page_setup, error))
-        return FALSE;
+	{
+	  unlink (filename);
+	  return FALSE;
+	}
     }
 
   /* The file will be removed when the GtkPrintJob closes it (once the job is
    * complete).
    */
-  unlink (filename);
+  if (!preview)
+    unlink (filename);
 
   return TRUE;
 }
@@ -504,8 +507,14 @@ handle_print (XdpImplPrint *object,
   gtk_window_set_transient_for (GTK_WINDOW (dialog), GTK_WINDOW (fake_parent));
   gtk_window_set_modal (GTK_WINDOW (dialog), modal);
   gtk_print_unix_dialog_set_manual_capabilities (GTK_PRINT_UNIX_DIALOG (dialog),
-                                                 can_preview () ? GTK_PRINT_CAPABILITY_PREVIEW : 0);
-
+                                                 can_preview () ? GTK_PRINT_CAPABILITY_PREVIEW : 0 |
+						 GTK_PRINT_CAPABILITY_PAGE_SET |
+						 GTK_PRINT_CAPABILITY_COPIES |
+						 GTK_PRINT_CAPABILITY_COLLATE |
+						 GTK_PRINT_CAPABILITY_REVERSE |
+						 GTK_PRINT_CAPABILITY_SCALE |
+						 GTK_PRINT_CAPABILITY_NUMBER_UP
+						 );
   handle = g_new0 (PrintDialogHandle, 1);
   handle->impl = object;
   handle->invocation = invocation;
@@ -652,7 +661,14 @@ handle_prepare_print (XdpImplPrint *object,
   gtk_window_set_transient_for (GTK_WINDOW (dialog), GTK_WINDOW (fake_parent));
   gtk_window_set_modal (GTK_WINDOW (dialog), modal);
   gtk_print_unix_dialog_set_manual_capabilities (GTK_PRINT_UNIX_DIALOG (dialog),
-                                                 can_preview () ? GTK_PRINT_CAPABILITY_PREVIEW : 0);
+                                                 can_preview () ? GTK_PRINT_CAPABILITY_PREVIEW : 0 |
+						 GTK_PRINT_CAPABILITY_PAGE_SET |
+						 GTK_PRINT_CAPABILITY_COPIES |
+						 GTK_PRINT_CAPABILITY_COLLATE |
+						 GTK_PRINT_CAPABILITY_REVERSE |
+						 GTK_PRINT_CAPABILITY_SCALE |
+						 GTK_PRINT_CAPABILITY_NUMBER_UP
+						 );
   gtk_print_unix_dialog_set_embed_page_setup (GTK_PRINT_UNIX_DIALOG (dialog), TRUE);
   gtk_print_unix_dialog_set_settings (GTK_PRINT_UNIX_DIALOG (dialog), settings);
   gtk_print_unix_dialog_set_page_setup (GTK_PRINT_UNIX_DIALOG (dialog), page_setup);
